// @elastic/charts/dist/chart_types/partition_chart/state/chart_state.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var __1 = require("../..");
var partition_1 = require("../renderer/canvas/partition");
var EMPTY_MAP = new Map();
var PartitionState = /** @class */ (function () {
    function PartitionState() {
        this.chartType = __1.ChartTypes.Partition;
    }
    PartitionState.prototype.isBrushAvailable = function () {
        return false;
    };
    PartitionState.prototype.isBrushing = function () {
        return false;
    };
    PartitionState.prototype.isChartEmpty = function () {
        return false;
    };
    PartitionState.prototype.getLegendItems = function () {
        return EMPTY_MAP;
    };
    PartitionState.prototype.getLegendItemsValues = function () {
        return EMPTY_MAP;
    };
    PartitionState.prototype.chartRenderer = function () {
        return react_1.default.createElement(partition_1.Partition, null);
    };
    PartitionState.prototype.getPointerCursor = function () {
        return 'default';
    };
    return PartitionState;
}());
exports.PartitionState = PartitionState;
