// @elastic/charts/dist/chart_types/partition_chart/renderer/canvas/partition.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var redux_1 = require("redux");
var react_redux_1 = require("react-redux");
var chart_1 = require("../../../../state/actions/chart");
var is_initialized_1 = require("../../../../state/selectors/is_initialized");
var geometries_1 = require("../../state/selectors/geometries");
var viewmodel_types_1 = require("../../layout/types/viewmodel_types");
var canvas_renderers_1 = require("./canvas_renderers");
var PartitionComponent = /** @class */ (function (_super) {
    __extends(PartitionComponent, _super);
    function PartitionComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.canvasRef = react_1.default.createRef();
        _this.ctx = null;
        _this.devicePixelRatio = window.devicePixelRatio;
        return _this;
    }
    PartitionComponent.prototype.drawCanvas = function () {
        if (this.ctx) {
            var _a = this.props.chartContainerDimensions, width = _a.width, height = _a.height;
            canvas_renderers_1.renderPartitionCanvas2d(this.ctx, this.devicePixelRatio, __assign(__assign({}, this.props.geometries), { config: __assign(__assign({}, this.props.geometries.config), { width: width, height: height }) }));
        }
    };
    PartitionComponent.prototype.tryCanvasContext = function () {
        var canvas = this.canvasRef.current;
        this.ctx = canvas && canvas.getContext('2d');
    };
    PartitionComponent.prototype.componentDidUpdate = function () {
        if (!this.ctx) {
            this.tryCanvasContext();
        }
        if (this.props.initialized) {
            this.drawCanvas();
            this.props.onChartRendered();
        }
    };
    PartitionComponent.prototype.componentDidMount = function () {
        // the DOM element has just been appended, and getContext('2d') is always non-null,
        // so we could use a couple of ! non-null assertions but no big plus
        this.tryCanvasContext();
        this.drawCanvas();
    };
    PartitionComponent.prototype.render = function () {
        var _a = this.props, initialized = _a.initialized, _b = _a.chartContainerDimensions, width = _b.width, height = _b.height;
        if (!initialized || width === 0 || height === 0) {
            return null;
        }
        return (react_1.default.createElement("canvas", { ref: this.canvasRef, className: "echCanvasRenderer", width: width * this.devicePixelRatio, height: height * this.devicePixelRatio, style: {
                width: width,
                height: height,
            } }));
    };
    PartitionComponent.displayName = 'Partition';
    return PartitionComponent;
}(react_1.default.Component));
var mapDispatchToProps = function (dispatch) {
    return redux_1.bindActionCreators({
        onChartRendered: chart_1.onChartRendered,
    }, dispatch);
};
var DEFAULT_PROPS = {
    initialized: false,
    geometries: viewmodel_types_1.nullSectorViewModel(),
    chartContainerDimensions: {
        width: 0,
        height: 0,
        left: 0,
        top: 0,
    },
};
var mapStateToProps = function (state) {
    if (!is_initialized_1.isInitialized(state)) {
        return DEFAULT_PROPS;
    }
    return {
        initialized: true,
        geometries: geometries_1.partitionGeometries(state),
        chartContainerDimensions: state.parentDimensions,
    };
};
exports.Partition = react_redux_1.connect(mapStateToProps, mapDispatchToProps)(PartitionComponent);
